### DISCUSSION TITLE
Does `@apply` suffer from the same downsides as `@extend` in SCSS?

### DISCUSSION THREAD

**Discussion by JoeyEamigh** [tailwindlabs/tailwindcss#6631] [Help] [answered]
This is a question I would love to see answered in the docs page under [https://tailwindcss.com/docs/functions-and-directives#apply](https://tailwindcss.com/docs/functions-and-directives#apply).

When using SCSS, using `@use` or `@import` in the SCSS file will include an entire copy of the file in the compiled SCSS. This means if I were to `@extend` a Bootstrap class per se, I would get an entire copy of Bootstrap in my SCSS file, leading to consistent redefines of variables, as well as massive compiled file sizes. 

Basically, I am asking if using `@apply` in my CSS or SCSS files to create a cleaner use of Tailwind unnecessarily convolutes the file size more than simply typing the classes into the HTML would?

I am trying to figure out there are massive downsides to using `@apply` to create cleaner HTML.

**Comment by adamwathan** [comment, marked as answer]
It doesn't work quite like `@extend` in Sass but it definitely adds weight to your CSS file. I've written a bit about the downsides of `@apply` in the documentation here:

https://tailwindcss.com/docs/reusing-styles#avoiding-premature-abstraction

I personally wouldn't recommend using it almost ever. It's the most complicated part of the whole framework and often doesn't do what you want it to do, even though it's doing the right thing.

If you don't want to use utility classes in your HTML, I would personally just recommend not using Tailwind. Being able to style things in your HTML instead of in a CSS file is the primary benefit, so if you aren't working that way you aren't really getting much out of using it that you couldn't get with just some variables to define some reusable values in your CSS.

**Comment by JoeyEamigh** [comment]
I decided to run a test with Create React App's default configuration (after following the documentation) to see how much size `@apply` adds to the bundle compared with `@use` paired with `@extend` in SCSS. This is mostly just a simple quantitative comparison for people like me who are merely curious of how the *not recommended* method of `@apply` compares to `@extend`.

This is a simple app with 3 pages: A main page, then two pages that are being imported with `React.lazy()` (since this is where a majority of the issues occur in SCSS).

With classes inlined:
```
File sizes after gzip:
  43.94 kB  build/static/js/main.0e96ca65.js
  1.46 kB   build/static/css/main.3ff83991.css
  291 B     build/static/js/880.bf5c9427.chunk.js
  280 B     build/static/js/576.a8ccac12.chunk.js
  20 B      build/static/css/576.31d6cfe0.chunk.css
  20 B      build/static/css/880.31d6cfe0.chunk.css
```

With classes extracted to SCSS file and `@apply`'d:
```
File sizes after gzip:
  43.91 kB (-24 B)    build/static/js/main.8e292109.js
  1.33 kB (+1.31 kB)  build/static/css/880.f46630d1.chunk.css
  1.16 kB (-304 B)    build/static/css/main.cc661936.css
  1.15 kB (+1.13 kB)  build/static/css/576.45808f4d.chunk.css
  260 B (-31 B)       build/static/js/880.1931a215.chunk.js
  259 B (-21 B)       build/static/js/576.07fefab7.chunk.js
  ```

The additions are not insignificant, but running this same test with Bootstrap 5 in its default configuration, and then using `@use` to import the configured `_variables.scss` (how bootstrap recommends set up) and `@extend` with a similar app ends with all CSS files having a minimum size of ~27kb unless you use a hacky plugin to extract classes (this rarely ends well).

To conclude, `@apply`, while not being the *recommended* way of using Tailwind, is much much better than using SCSS `@use` and `@extend` to achieve the same effect, meaning there are perceivable benefits of using Tailwind for devs who like the `@extend` pattern.

This was just a fun little project I did to appease my curiosity at 2am. Thank you for coming to my TED Talk, and thank you adamwathan and other Tailwind maintainers for your work!
